field/TinyLang/Field/Typed/TypeChecker.hs

```diff
@@ -186,71 +186,65 @@ withTypedBinOp R.BAt k = k T.BAt
--}
-checkUniVar
-    :: forall m f a. (MonadTypeChecker m)
-    => Uni f a -> R.Var -> m (T.UniVar f a)
-checkUniVar uni iden = do
-    Some uniVar@(T.UniVar varUni _) <- inferUniVar iden
-    let uniMismatch = typeMismatch uniVar uni varUni
-    withGeqUniM uni varUni uniMismatch uniVar
-
 {-| Type checking for expressions
 -}
-checkExpr
-    :: forall m f a. (MonadTypeChecker m, TextField f, KnownUni f a)
-    => R.Expr R.Var f -> m (T.Expr f a)
+checkExpr ::
+    forall m f a. (Monad m, TextField f, KnownUni f a)
+    => R.Expr R.Var f -> TypeCheckerT m (T.Expr f a)
 checkExpr (R.EIf l m n) = T.EIf <$> checkExpr l <*> checkExpr m <*> checkExpr n
 checkExpr m = do
     SomeOf mUni tM <- inferExpr m
     let uni = knownUni @f @a
     let uniMismatch = typeMismatch tM uni mUni
     withGeqUniM uni mUni uniMismatch tM
 
--- TODO:  Figure how to optimise
-checkProgram
-    :: forall m f. (MonadTypeChecker m, TextField f)
-    => R.Program R.Var f -> m (T.Program f)
-checkProgram (R.Program (R.Statements stmts)) =
-    T.Program . T.Statements <$> foldMapA checkStatement stmts
+checkProgram ::
+    forall m f. (Monad m, TextField f)
+    => R.Program R.Var f -> TypeCheckerT m (T.Program f)
+checkProgram (R.Program stmts) =
+    checkStatements stmts $ pure . T.Program
 
+checkStatements ::
+    forall m f. (Monad m , TextField f)
+    => R.Statements R.Var f -> forall r. (T.Statements f -> TypeCheckerT m r) -> TypeCheckerT m r
+checkStatements (R.Statements stmts) kont =
+    runContT (foldMapA (ContT . checkStatement) stmts) $ kont . T.Statements
 
 {-| Type checking judgement for statements of form
 -}
-checkStatement
-    :: forall m f. (MonadTypeChecker m, TextField f)
-    => R.Statement R.Var f -> m [T.Statement f]
-checkStatement (R.ELet var m) = do
+checkStatement ::
+    forall m f. (Monad m , TextField f)
+    => R.Statement R.Var f -> forall r. ([T.Statement f] -> TypeCheckerT m r) -> TypeCheckerT m r
+checkStatement (R.ELet var m) kont = do
     Some uniVar@(T.UniVar uni _) <- inferUniVar var
-    T.withKnownUni uni $ (:[]) . T.ELet uniVar <$> checkExpr m
-checkStatement (R.EAssert m) =
-    (:[]) . T.EAssert <$> checkExpr m
--- TODO:  Inject var into environment of stmts
-checkStatement (R.EFor var start end stmts) =
-    foldMapA iter [start .. end]
-    where iter i = do
-              tVar <- makeVar $ R.unVar var
-              let uVar = T.UniVar Field tVar
-              tStmts <- foldMapA checkStatement stmts
-              pure $ T.ELet uVar (T.EConst . fromIntegral $ i) : tStmts
+    tM <- T.withKnownUni uni $ checkExpr m
+    withVar var $ kont [T.ELet uniVar tM]
+checkStatement (R.EAssert m) kont = do
+    tM <- checkExpr m
+    kont [T.EAssert tM]
+checkStatement (R.EFor var start end stmts) kont = do
+    T.Statements tStmts <- withVar var $ checkStatements stmts pure
+    tVar <- makeVar $ R.unVar var
+    let uVar = T.UniVar Field tVar
+    let iter i ikont = ikont $ T.ELet uVar (T.EConst . fromIntegral $ i) : tStmts
```

Comment:
So in the typed language `var` is in scope after the loop?

Reply:
Yes and No.

Yes, because we unroll a loop to a series of let bindings, meaning that technically every variable introduced is in scope.

No, because it inaccessible after the loop.  The following program will not type check
```
for x = 0 to 0 do; end;
assert x == 0;
```

Reply:
> No, because it inaccessible after the loop. The following program will not type check
> ```
> for x = 0 to 0 do; end;
> assert x == 0;
> ```

Hmm, okay. What does

```
let x = 1;
for x = 0 to 0 do; end;
assert x == 0;
```

elaborate to? Looking at the current version with `withVar` I feel like this would elaborate to the wrong thing, but not sure. Could you add such a golden test?

Reply:
I am resolving this conversation as we went back to the previous notion of loops, where the loop variable is bound after the loop.

Reply:
> I am resolving this conversation as we went back to the previous notion of loops, where the loop variable is bound after the loop.

Is this something that we're going to have in the long run? If the plan is to inline the value of a loop variable, then we can unscope the loop variable after the loop (and only inline in the body of the loop then).